core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -220,62 +218,65 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     const Kokkos::Experimental::SYCL& space = m_policy.space();
     Kokkos::Experimental::Impl::SYCLInternal& instance =
         *space.impl_internal_space_instance();
-    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMemory& kernelMem =
-        *instance.m_indirectKernel;
-
-    // Allocate USM shared memory for the functor
-    kernelMem.resize(std::max(kernelMem.size(), sizeof(functor)));
+    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMem&
+        indirectKernelMem = instance.m_indirectKernelMem;
 
-    // Placement new a copy of functor into USM shared memory
-    //
-    // Store it in a unique_ptr to call its destructor on scope exit
-    std::unique_ptr<Functor, Kokkos::Impl::destruct_delete> kernelFunctorPtr(
-        new (kernelMem.data()) Functor(functor));
+    // Store a copy of the functor in indirectKernelMem
+    auto kernelFunctorPtr = indirectKernelMem.copy_from(functor);
 
     auto kernelFunctor = ExtendedReferenceWrapper<Functor>(*kernelFunctorPtr);
     sycl_direct_launch(m_policy, kernelFunctor);
   }
 
+  void zero_length_reduction() const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = m_policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    sycl::queue& q = *instance.m_queue;
+    Kokkos::Experimental::Impl::SYCLInternal::ReductionResultMem&
+        reductionResultMem = instance.m_reductionResultMem;
+
+    // reductionResultMem - USM shared memory
+    // fancy_result_ptr - manages object (if any) in reductionResultMem
+    // result_ptr - raw pointer (either m_result_ptr if accessible on
+    // both the host and device OR to reductionResultMem)
+
+    value_type host_result;
+    decltype(reductionResultMem.copy_from(host_result)) fancy_result_ptr;
+    pointer_type result_ptr;
+
+    const sycl::usm::alloc result_ptr_type =
+        sycl::get_pointer_type(m_result_ptr, q.get_context());
+
+    switch (result_ptr_type) {
+      case sycl::usm::alloc::host:
+      case sycl::usm::alloc::shared:
+        // We can access m_result_ptr both on the host and device
+        ValueInit::init(m_functor, m_result_ptr);
+        result_ptr = m_result_ptr;
+        break;
+      case sycl::usm::alloc::device:
+      case sycl::usm::alloc::unknown:  // non-USM-allocated memory
+        // We can't access m_result_ptr both on the host and device
+        ValueInit::init(m_functor, &host_result);
+        fancy_result_ptr = reductionResultMem.copy_from(host_result);
+        result_ptr       = &*fancy_result_ptr;
```

Comment:
This looks rather weird since `fancy_result_ptr` can either be a raw pointer or a `std::unique_ptr`.

Reply:
`copy_from` now always returns a `unique_ptr`.